File: codiffs.pl
#
# File: codiffs.pl
# Author: eweb
# Copyright WBT Systems, 2003-2011
# Contents: Perl script to do diffs of cheked out files
#
# Date:          Author:  Comments:
# 21st Sep 2007  eweb     #00008 Adapted for p4 and escc.
#  8th Apr 2009  eweb     #00008 Handle existing comments.
# 16th Apr 2009  eweb     #00008 Quote filenames
# 28th May 2009  eweb     #00008 p4 & git
#  2nd Nov 2009  eweb     #00008 git before escc
#  4th Dec 2009  eweb     #00008 Tidy up
# 18th Mar 2010  eweb     #00008 Subversion
# 29th Mar 2010  eweb     #00008 Better handling of multi line comments
#  6th May 2010  eweb     #00008 Don't output auto directory comments
# 31st May 2010  eweb     #00008 Missing last checkout if it has multi-line comment
#  8th Aug 2010  eweb     #00008 Options, first determine scc then set cmds etc
# 19th Aug 2010  eweb     #00008 git options
#  6th Jan 2011  eweb     #00008 Name the identical file
#  7th Feb 2011  eweb     #00008 Handle git new files
# 29th Mar 2011  eweb     #00008 -f filter
#

use strict;
use Cwd;
use Getopt::Std;

my $cctool1 = "cleartool"; # info gathering
my $cctool2 = "cleartool"; # easily reversable
my $cctool3 = "cleartool"; # reversable
my $cctool4 = "cleartool"; # destructive

my $lsco = "$cctool1 lsco -cview -avobs -fmt \"%n [%Nc]\\n\"";
my $diff1 = "$cctool1 diff -pred -serial";
my $diff2 = "$cctool1 diff -pred -serial -options -b";
#$diff1 = $diff2;
my $namepatt = '^(.*) \[(.*)\]$';
my $namepatt2;
my $skipline;
#$namepatt = '^(.*topclass.js)$';

my $listIdents = "N";
#$listIdents = "Y";
my $verbose;
#$verbose = 1;
my $nodiffs;
#$nodiffs = 1;
my $startline;
my $stopline;
my $filter;
my $substPattern;
my $substSubst;
my $editor = "textpad";

my $cwd = getcwd();

print "cwd: $cwd\n" if ($verbose);

my %opts = ( v => undef(),
             o => undef(),
             s => undef(),
             b => undef(),
             f => undef(),
           );

if ( !getopts("v:o:s:bf:", \%opts) or @ARGV > 1 ) {
  print STDERR "Unknown arg $ARGV[0]\n" if @ARGV > 0;
  #Usage();
  exit;
}

my $ignorespace = $opts{b};
my $verbose = $opts{v};
my $scc = $opts{s};
my $options = $opts{o};
my $filter = $opts{f};

$scc = "clearcase" unless ( $scc );

my $output;
my $cosout;

my $drive = "c";
if ( $cwd =~ /^([a-z]):/i ) {
    $drive = lc $1;
}

if ( $cwd =~ /^c:\/p4clients/i or $cwd =~ /^c:\\p4clients/i ) {
    $scc = "p4";
}
elsif ( -d "\\.git" or -d ".git" ) {
    $scc = "git";
}
elsif ( -d "\\.svn" or -d ".svn" ) {
    $scc = "svn";
}
elsif ( $cwd =~ /^c:\/cpp/i or $cwd =~ /^c:\\cpp/i ) {
    $scc = "escc";
}
else {
}

if ( $scc eq "clearcase" ) {
    print "clearcase\n" if ($verbose);
    my $lsco = "$cctool1 lsco -cview -avobs -fmt \"%n [%Nc]\\n\"";
    my $diff1 = "$cctool1 diff -pred -serial";
    my $diff2 = "$cctool1 diff -pred -serial -options -b";
    #$diff1 = $diff2;
    my $namepatt = '^(.*) \[(.*)\]$';
    my $skipline;
}
elsif ( $scc eq "p4" ) {
    print "p4\n" if ($verbose);
    $lsco = "p4 diff -sa";
    $lsco = "p4 opened"; #//depot/wacc/java/acc/test/org/eweb/cpp/CalcTest.java#1 - add change 201 (text)
    $diff1 = "p4 diff";
    $diff2 = "p4 diff -dbwc";
    $namepatt = "^(.+)#[0-9]+ - .+";
    $substPattern = "//depot/";
    $substSubst = "c:/p4clients/floyd/";
    $drive = "p";
}
elsif ( $scc eq "git" ) {
    print "git\n" if ($verbose);
    $lsco = "git status";
    if ( $options eq "all" ) {
      # working tree relative to the named commit (HEAD)
      $diff1 = "git diff HEAD";
      $diff2 = "git diff -w HEAD";
    }
    elsif ( $options eq "cached" ) {
      # staged relative to the named commit (HEAD)
      $diff1 = "git diff --cached";
      $diff2 = "git diff -w --cached";
    }
    else {
      # changes relative to the index
      $diff1 = "git diff";
      $diff2 = "git diff -w";
      $startline = "Changed but not updated";
    }
    $namepatt = "^#\\s+modified:\\s+(.+)";
    $namepatt2 = "^#\\s+new file:\\s+(.+)";
    $skipline = "^#";
    if ( -d "\\.git" ) {
      chdir( "\\" );
    }
}
elsif ( $scc eq "svn" ) {
    print "svn\n" if ($verbose);
    $lsco = "svn status -q";
    $diff1 = "svn diff";
    $diff2 = "svn diff";
    $namepatt = "^[AM].......(.+)";
    $skipline = "^[^AM]";
}
elsif ( $scc eq "escc" ) {
    print "escc\n" if ($verbose);
    $lsco = "escc lsco";
    $diff1 = "escc diff";
    $diff2 = "escc diff";
    $namepatt = "^([^[]+)";
}
else {
    print "unknown scc $scc\n";
}

if ( $output eq "" ) {
    if ( $^O eq "darwin" ) {
      $output = $ENV{HOME} . "/$drive-codiffs.txt";
      $cosout = $ENV{HOME} . "/$drive-cos.sh";
      $editor = "emacs";
    }
    else {
      $output = "c:\\temp\\$drive-codiffs.txt";
      $cosout = "c:\\temp\\$drive-cos.bat";
    }
    print " $editor $output $cosout\n";
}

if ( $output eq "" ) {
    $output = "-";
}

open( OUT, ">$output" );

if ( $cosout ne "" ) {
    open( COSOUT, ">$cosout" );
}

my $addcomment = "addcomment.pl";

if ( open( COS, "$lsco 2>&1 |" ) ) {
    my $skip;
    if ( $startline ne "" ) {
      $skip = 1;
    }
    my $c = 0;
    while ( <COS> ) {
        $c++;
        print "$c: $_" if ( $verbose );
        chomp;
        my $co = $_;
        my $file;
        my $comments;
        if ( $skip ) {
            if ( $co =~ /$startline/ ) {
                $skip = undef;
            }
            else {
                next;
            }
        }
        if ( $substPattern && $substSubst ) {
          $co =~ s!\Q$substPattern\E!$substSubst!;
        }
        if ( $co =~ /$namepatt/s ) {
            $file = $1;
            $comments = $2;
            print "\$file: $file \$comments: $comments\n" if ( $verbose );
        }
        elsif ( $namepatt2 and $co =~ /$namepatt2/s ) {
            $file = $1;
            $comments = $2;
            print "\$file: $file \$comments: $comments\n" if ( $verbose );
        }
        elsif ( $skipline and $co =~ /$skipline/ ) {
            print "Skipping line\n" if ( $verbose );
        }
        else {
            print "continuing line\n" if ( $verbose );
            unless ( eof(COS) ) {
              $_ .= "\n" . <COS>; # separate with a line break
              redo; # unless eof(COS);
            }
        }
        print "checking file $file\n" if ( $verbose );
        if ( $file && -d $file ) {
            # need to extract the non automatic comments...
            # split comment into lines, remove all the auto adding element, uncataloging symlink, etc.
        }
        if ( $filter && $file !~ /$filter/ ) {
          next;
        }
        if ( $comments ) {
            # for each line of comments...
            my @clines = split( /[\n\r]+/, $comments );

            if ( $comments !~ /^#/ ) {
                print COSOUT "    $addcomment -c \"\" -insert \"$file\"\n";
            }
            foreach ( @clines ) {
                s!"!\\"!g;
                if ( -d $file and ( /Added file element/ or /Added directory element/ or /Added symbolic link/ or /Uncataloged directory element/ or /Uncataloged file element/ ) ) {
                    #print COSOUT "REM $addcomment -c \"$_\" \"$file\"\n";
                }
                else {
                    print COSOUT "REM $addcomment -c \"$_\" \"$file\"\n";
                }
            }
        }
        else {
            if ( -d $file ) {
                print COSOUT "    $addcomment -c \"\" -insert \"$file\"\n";
            }
            elsif ( -e $file ) {
                print COSOUT "    $addcomment -c \"\" \"$file\"\n";
            }
        }
        if ( $nodiffs ) {
        }
        elsif ( $file ) {
            my $cmd = "$diff1 \"$file\"";
            if ( $ignorespace ) {
                $cmd = "$diff2 \"$file\"";
            }
            print "$cmd\n" if ( $verbose );
            if ( open( DIFF, "$cmd 2>&1 |" ) ) {
                while ( <DIFF> ) {
                    s/[\r\n]+$//;
                    if ( /^Files are identical$/ or /^Directories are identical$/ ) {
                        print OUT "$file: $_\n";
                        if ( $listIdents eq "Y" ) {
                            print OUT "$file\n";
                            CheckIn( $file, "" );
                        }
                        else {
                            last;
                        }
                    }
                    elsif ( /^old mode/ or /^new mode/ ) {
                    }
                    else {
                        if ( $_ ne "" ) {
                            print OUT "$_\n";
                        }                        
                    }
                }
                close( DIFF );
            }
        }
    }
    close( COS );
}

if ( $listIdents eq "Y" and $diff1 ne $diff2 ) {
    if ( open( COS, "$lsco 2>&1 |" ) ) {
        while ( <COS> ) {
            chomp;
            if ( /$namepatt/ ) {
                my $file = $1;
                if ( $nodiffs ) {
                }
                elsif ( open( DIFF, "$diff2 $file 2>&1 |" ) ) {
                    while ( <DIFF> ) {
                        if ( $listIdents eq "Y" ) {
                            if ( /^Files are identical$/ ) {
                                print OUT "$file\n";
                                CheckIn( $file, "#????? White space" );
                            }
                            else {
                                last;
                            }
                        }
                        else {
                            print OUT $_;
                        }
                    }
                    close( DIFF );
                }
            }
        }
        close( COS );
    }
}

close( OUT );
close( COSOUT );

sub CheckIn($$) {
    my ($file, $comment) = @_;

    my $cmd;
    if ( $comment ne "" ) {
        $cmd = "$cctool2 ci -c \"$comment\" $file";
    }
    else {
        $cmd = "$cctool2 ci -nc $file";
    }
    print OUT "$cmd\n";
    my $results = `$cmd 2>&1`;

    if ( $results =~ /Error: By default, won't create version with data identical to predecessor./ ) {
        # hasn't changed so undo the check out.
        $cmd = "$cctool4 unco -rm $file";
        $results = $results . "\n" . $cmd . "\n" . `$cmd 2>&1`;
    }
    elsif ( $results =~ /Error: Not an element:/ ) {
        # Not an element
    }
    elsif ( $results =~ /Error:/ ) {
        # Not an element
    }
    else {
        # Not an element
    }
    print OUT "$results\n";
}

File: codiffs.rb
#!/usr/bin/env ruby
#
# File: codiffs.rb
# Author:
# Copyright eweb, 2012-2015
# Contents:
#
# Date:          Author:  Comments:
# 19th Jul 2012  eweb     #0008 Initial port to ruby
#  5th Aug 2012  eweb     #0008 Write codiffs.txt
#  5th Aug 2012  eweb     #0008 Include git add commands
#  5th Aug 2012  eweb     #0008 git add only for untracked files
#  6th Aug 2012  eweb     #0008 Make cos.sh executable
# 10th Sep 2012  eweb     #0008 deleted, modified, renamed and new file
# 10th Sep 2012  eweb     #0008 pushd and popd
# 10th Sep 2012  eweb     #0008 make cos.sh executable
# 24th Oct 2012  eweb     #0008 Report number of files changed
#  7th Nov 2013  eweb     #0008 Use addcomment.rb
# 24th Jun 2014  eweb     #0008 git output changes, no diffs file
# 20th Aug 2014  eweb     #0008 Cleaner output
#  3rd Sep 2014  eweb     #0008 Delete file if no changes
# 20th Sep 2014  eweb     #0008 Don't delete cos.sh
# 25th Mar 2015  eweb     #0008 Always push and pop dir
#

def find_git( where = "." )
  where = File.expand_path where
  #puts "Trying #{where}\n"
  if File.directory?("#{where}/.git")
    where
  elsif where == '/'
  else
    up = File.expand_path "#{where}/.."
    #puts "up #{up}\n"
    find_git up if up != where
  end
end

verbose = false

project_root = find_git

exit unless project_root

addcomment = 'addcomment.rb'

need_to_change_directory = Dir.pwd != project_root

Dir.chdir project_root if need_to_change_directory

script_file = project_root + '/cos.sh'
#diffs_file = project_root + '/codiffs.txt'

#puts script_file

changed_files = []

  IO.popen("git status") do |f|
    stage = nil
    f.each do |line|
      #puts line
      if line =~ /On branch (.*)/
        branch = $1
      elsif line =~ /Your branch is up-to-date with '.+'/
      elsif line =~ /Your branch is ahead of '.+'/
      elsif line =~ /nothing to commit, working directory clean/
      elsif line =~ /Changes to be committed:/
        stage = :staged
      elsif line =~ /\tdeleted: +(.*)/
        #
      elsif line =~ /\tmodified: +(.*)/ ||
            line =~ /\trenamed: +(?:.*) -> (.*)/ ||
            line =~ /\tnew file: +(.*)/
        file = $1
        puts "Staged #{file}\n" if stage == :staged && verbose
        puts "Unstaged #{file}\n" if stage == :unstaged && verbose
        changed_files << stage
        changed_files << file
      elsif line =~ /Untracked files:/
        stage = :untracked
      elsif line =~ /Changes not staged for commit:/
        stage = :unstaged
      elsif line =~ /\t(.*)/
        file = $1
        puts "Untracked #{file}" if verbose
        changed_files << stage
        changed_files << file
      elsif line =~ / \(use/
      elsif line == "\n"
      elsif line =~ /^[^#]/
        puts "Unhandled #{line}"
      else
        puts "Unhandled #{line}"
      end
    end
  end

def get_comments file #, diffs
  comments = []
  # git diff compares working tree with index what could be staged
  # git diff --cached compares index with HEAD what has been staged
  IO.popen("git diff HEAD -- #{file}") do |f|
    f.each do |line|
      if line =~ /\+# ([0-9]{1,2}th|st|nd|rd) ([A-Z][a-z]{2}) ([0-9]{4})  (.+)     (.+)/
        puts "Comment: #{$1} #{$2} #{$3} #{$4} #{$5}\n"
        comments << $5
      else
        #diffs.puts line
      end
    end
  end
  comments
rescue => e
  puts "Error scanning file #{file} for comments #{e}"
  []
end

if changed_files.length
  files_changed = 0
  File.open( script_file, "w" ) do |script|
    script.puts "pushd #{project_root}" # if need_to_change_directory
    stage =
    changed_files.each do |f|
      if f.class == Symbol
        script.puts "### #{f}" if stage != f
        stage = f
      else
        files_changed = files_changed + 1
        comments = get_comments f #, diffs
        if stage == :untracked
          script.puts "#git add #{f}"
        end
        if comments.length > 0
          comments.each do |c|
            script.puts "##{addcomment} -c \"#{c}\" #{f}"
          end
        else
          script.puts "#{addcomment} -c \"\" #{f}"
        end
      end
    end
    script.puts "popd" # if need_to_change_directory
  end
  puts "#{files_changed} files changed"

  mode = File.stat( script_file ).mode & 0777
  # executable by owner
  File.chmod( mode | 0100, script_file )
  if files_changed > 0
    puts "aquamacs #{script_file}"
  end
end

